File: src/services/parser/v2/diff.ts
import { Group, Groups, Teacher, Teachers } from "../types";

type DiffLine = {
    key: string,
    day: string,
    reason: string
}

function diffEntries(newEntry: Group | Teacher, oldEntry?: Group | Teacher): DiffLine[] {
    const lines: DiffLine[] = [];
    const oldMap = new Map<string, Group['days'][number] | Teacher['days'][number]>();

    if (oldEntry) {
        for (const day of oldEntry.days) {
            oldMap.set(day.day, day);
        }
    }

    for (const day of newEntry.days) {
        const oldDay = oldMap.get(day.day);
        if (!oldDay) {
            lines.push({ key: 'new', day: day.day, reason: 'added' });
            continue;
        }

        if (JSON.stringify(oldDay.lessons) !== JSON.stringify(day.lessons)) {
            lines.push({ key: 'changed', day: day.day, reason: 'updated' });
        }

        oldMap.delete(day.day);
    }

    for (const [day] of oldMap) {
        lines.push({ key: 'old', day: day, reason: 'removed' });
    }

    return lines;
}

export function diffGroups(current: Groups, previous: Groups, limit: number): string[] {
    const output: string[] = [];
    const keys = new Set<string>([...Object.keys(current), ...Object.keys(previous)]);

    for (const key of keys) {
        const lines = diffEntries(current[key], previous[key]);
        for (const line of lines) {
            output.push(`${key}: ${line.reason} ${line.day}`);
            if (output.length >= limit) {
                return output;
            }
        }
    }

    return output;
}

export function diffTeachers(current: Teachers, previous: Teachers, limit: number): string[] {
    const output: string[] = [];
    const keys = new Set<string>([...Object.keys(current), ...Object.keys(previous)]);

    for (const key of keys) {
        const lines = diffEntries(current[key], previous[key]);
        for (const line of lines) {
            output.push(`${key}: ${line.reason} ${line.day}`);
            if (output.length >= limit) {
                return output;
            }
        }
    }

    return output;
}

File: internal/parser/v2/diff.go
package v2

import (
	"encoding/json"

	"github.com/blindmaster24/MgkeTimetableBot/internal/model"
)

type DiffLine struct {
	Key    string
	Day    string
	Reason string
}

func DiffGroups(current, previous model.Groups, limit int) []string {
	return diffAll(current, previous, limit)
}

func DiffTeachers(current, previous model.Teachers, limit int) []string {
	return diffAll(current, previous, limit)
}

func diffAll[T model.Group | model.Teacher](current, previous map[string]*T, limit int) []string {
	keys := make(map[string]bool)
	for k := range current {
		keys[k] = true
	}
	for k := range previous {
		keys[k] = true
	}

	var output []string
	for key := range keys {
		lines := diffEntries(current[key], previous[key])
		for _, line := range lines {
			output = append(output, key+": "+line.Reason+" "+line.Day)
			if len(output) >= limit {
				return output
			}
		}
	}

	return output
}

func diffEntries[T model.Group | model.Teacher](newEntry, oldEntry *T) []DiffLine {
	var lines []DiffLine

	var newDays []struct{ Day string; Lessons interface{} }
	var oldDays []struct{ Day string; Lessons interface{} }

	switch e := any(newEntry).(type) {
	case *model.Group:
		if e == nil {
			break
		}
		for _, d := range e.Days {
			newDays = append(newDays, struct{ Day string; Lessons interface{} }{d.Day, d.Lessons})
		}
	case *model.Teacher:
		if e == nil {
			break
		}
		for _, d := range e.Days {
			newDays = append(newDays, struct{ Day string; Lessons interface{} }{d.Day, d.Lessons})
		}
	}

	if oldEntry != nil {
		switch e := any(oldEntry).(type) {
		case *model.Group:
			if e == nil {
				break
			}
			for _, d := range e.Days {
				oldDays = append(oldDays, struct{ Day string; Lessons interface{} }{d.Day, d.Lessons})
			}
		case *model.Teacher:
			if e == nil {
				break
			}
			for _, d := range e.Days {
				oldDays = append(oldDays, struct{ Day string; Lessons interface{} }{d.Day, d.Lessons})
			}
		}
	}

	oldMap := make(map[string]interface{})
	for _, d := range oldDays {
		oldMap[d.Day] = d.Lessons
	}

	for _, d := range newDays {
		oldLessons, exists := oldMap[d.Day]
		if !exists {
			lines = append(lines, DiffLine{Key: "new", Day: d.Day, Reason: "added"})
			continue
		}
		newJSON, _ := json.Marshal(d.Lessons)
		oldJSON, _ := json.Marshal(oldLessons)
		if string(newJSON) != string(oldJSON) {
			lines = append(lines, DiffLine{Key: "changed", Day: d.Day, Reason: "updated"})
		}
		delete(oldMap, d.Day)
	}

	for day := range oldMap {
		lines = append(lines, DiffLine{Key: "old", Day: day, Reason: "removed"})
	}

	return lines
}
